### ISSUE TITLE
Add BPM field for songs

### ISSUE THREAD

**Issue by kateabr** [VocaDB/vocadb#558]
A suggestion: field for storing songs' tempo.
Some producers provide it for their songs themselves, or it can be extracted from ust/midi/vsq etc. Being able to look the tempo up can save cover artists lots of time and nerve: trying to adjust voice synth to fit the instrumental's tempo is usually a pain (at least, for me).
It could also be a useful searching parameter for browsing faster/slower songs.

**Comment by riipah** [comment]
Is variable BPM relevant? There's definitely some songs like that, usually progressive rock / metal. Easiest way to handle variable BPM is to provide only one average BPM value. A range would also be possible, but it would be more complicated and would make sorting impossible.

This task only includes the field. Filtering and sorting can be added later.

**Comment by riipah** [comment]
[Discussion thread](https://vocadb.net/discussion/topics/268)

**Comment by andreoda** [comment]
For example iTunes has only one BPM field.
I am using BeaTunes to analyse my music library. It uses two fields. One for the one that it analysed and the second one is an alternate BPM.
Anyway, I think that one field is enough. Usually songs have only one BPM as its base.

**Comment by ycanardeau** [comment]
I think having minimum and maximum BPM values is the best compromise for the purpose of searching (just like event's start and end dates). For example, the BPM of [this song](https://vocadb.net/S/81641) varies from 94 to 160, but it doesn't make sense to have only one average or median BPM value.

**Comment by dset0x** [comment]
I'm a little late here! I see there's a minBpm and maxBpm field in the commit. I am not sure if these are the most useful to have.

In the vocadb discussion I see that there is usefulness in the BPM of (a) the beginning and (b) ending of the song (for mixing purposes). I also think it's good to have a (c) "Base (average) BPM". A (d) "secondary" BPM might also be useful. You can also go all out (like you mentioned in another issue) and have an arbitrarily long list of BPM segments.

While minimum and maximum can also be useful, they can also be deceiving (a 5 minute song might have a slow 10 second introduction). We could keep them, but perhaps not on their own.

Another important thing is to be able to detect whichever values we need. It would be interesting for me to see if I can put something together to detect these values and then auto-update songs.

**Comment by riipah** [comment]
I don't know about min-max BPM vs. average BPM. But "arbitrarily long list of BPM segments" sounds overkill for VocaDB. I suggested earlier that detailed BPM is complicated enough to justify having it on a separate website (linked to VocaDB). So if we're going to have a BPM value on VocaDB, it should be kept as simple as possible, 1 or 2 values.

**Comment by dset0x** [comment]
Sure, that's overkill (unless we really do want it for some reason and just provide it over API but don't render all the details on the Song page). Unfortunately I don't see an auxiliary page existing like that with any popularity.

**Comment by ycanardeau** [comment]
I also read the discussion on VocaDB, but I had to make a compromise. The most elegant part of min-max BPM is that it can be reused for songs that have no tempo changes. The current implementation regards "BPM range" as "Specific BPM", if and only if `MinBpm` and `MaxBpm` have the same value. If we introduced `BeginningBpm` and `EndingBpm`, we would have to add yet another column for "Specific BPM".
